internal/internal_nexus_task_handler.go

```diff
@@ -416,3 +422,93 @@ func (p *payloadSerializer) Serialize(v any) (*nexus.Content, error) {
+	Status() *status.Status
+}
+
+// convertServiceError converts a serviceerror into a Nexus HandlerError if possible.
+// If exposeDetails is true, the error message from the given error is exposed in the converted HandlerError, otherwise,
+// a default message with minimal information is attached to the returned error.
+// Roughly taken from https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto
+// and
+// https://github.com/grpc-ecosystem/grpc-gateway/blob/a7cf811e6ffabeaddcfb4ff65602c12671ff326e/runtime/errors.go#L56.
+func convertServiceError(err error, exposeDetails bool) error {
+	var st *status.Status
+	var stGetter statusGetter
+	if !errors.As(err, &stGetter) {
+		// Not a serviceerror, passthrough.
+		return err
+	}
+
+	st = stGetter.Status()
+	errMessage := err.Error()
+
+	switch st.Code() {
+	case codes.AlreadyExists, codes.Canceled, codes.InvalidArgument, codes.FailedPrecondition, codes.OutOfRange:
+		if !exposeDetails {
+			errMessage = "bad request"
+		}
+		return nexus.HandlerErrorf(nexus.HandlerErrorTypeBadRequest, errMessage)
+	case codes.Aborted, codes.Unavailable:
+		if !exposeDetails {
+			errMessage = "service unavailable"
+		}
+		return nexus.HandlerErrorf(nexus.HandlerErrorTypeUnavailable, errMessage)
+	case codes.DataLoss, codes.Internal, codes.Unknown:
+		if !exposeDetails {
+			errMessage = "internal error"
+		}
+		return nexus.HandlerErrorf(nexus.HandlerErrorTypeInternal, errMessage)
+	case codes.Unauthenticated:
+		if !exposeDetails {
+			errMessage = "authentication failed"
+		}
+		return nexus.HandlerErrorf(nexus.HandlerErrorTypeUnauthenticated, errMessage)
+	case codes.PermissionDenied:
+		if !exposeDetails {
+			errMessage = "permission denied"
+		}
+		return nexus.HandlerErrorf(nexus.HandlerErrorTypeUnauthorized, errMessage)
+	case codes.NotFound:
+		if !exposeDetails {
+			errMessage = "not found"
+		}
+		return nexus.HandlerErrorf(nexus.HandlerErrorTypeNotFound, errMessage)
+	case codes.ResourceExhausted:
+		if !exposeDetails {
+			errMessage = "resource exhausted"
+		}
+		return nexus.HandlerErrorf(nexus.HandlerErrorTypeResourceExhausted, errMessage)
+	case codes.Unimplemented:
+		if !exposeDetails {
+			errMessage = "not implemented"
+		}
+		return nexus.HandlerErrorf(nexus.HandlerErrorTypeNotImplemented, errMessage)
+	case codes.DeadlineExceeded:
+		if !exposeDetails {
+			errMessage = "request timeout"
+		}
+		return nexus.HandlerErrorf(nexus.HandlerErrorTypeDownstreamTimeout, errMessage)
+	}
+
+	// Default to internal error. This should only happen for codes.OK, which is unexpected for serviceerrors.
+	if !exposeDetails {
+		return nexus.HandlerErrorf(nexus.HandlerErrorTypeInternal, "internal error")
+	}
+	return err
+}
+
+// convertApplicationError converts a Temporal ApplicationError to a Nexus HandlerError, respecting the non_retryable
+// flag.
+func convertApplicationError(err error) error {
+	var appErr *ApplicationError
+	if errors.As(err, &appErr) {
```

Comment:
Normally the Go SDK explicitly does does not check the error chain for `ApplicationError`

Reply:
Same with the Server

Reply:
Changed this.